Sources/BraveWallet/WalletAssetBalanceManager.swift

```diff
@@ -0,0 +1,121 @@
+ 
+  private let keyringService: BraveWalletKeyringService
+  private let rpcService: BraveWalletJsonRpcService
+  private let assetManager: WalletUserAssetManager
+  
+  init(
+    keyringService: BraveWalletKeyringService,
+    rpcService: BraveWalletJsonRpcService,
+    assetManager: WalletUserAssetManager
+  ) {
+    self.keyringService = keyringService
+    self.rpcService = rpcService
+    self.assetManager = assetManager
+  }
+  
+  public func getBalance(for asset: BraveWallet.BlockchainToken?, account: String?) -> [WalletUserAssetBalance]? {
+    WalletUserAssetBalance.getBalance(for: asset, account: account)
+  }
+  
+  public func updateBalance(for asset: BraveWallet.BlockchainToken, account: String, balance: String, completion: (() -> Void)?) {
+    WalletUserAssetBalance.updateBalance(
+      for: asset,
+      balance: balance,
+      account: account,
+      completion: completion
+    )
+  }
+  
+  public func removeBalance(for asset: BraveWallet.BlockchainToken, completion: (() -> Void)?) {
+    WalletUserAssetBalance.removeBalance(
+      for: asset,
+      completion: completion
+    )
+  }
+  
+  public func refreshBalances() async {
+    let accounts = await keyringService.allAccounts().accounts
+    let allNetworks = await rpcService.allNetworksForSupportedCoins()
+    let allUserAssets: [NetworkAssets] = assetManager.getAllUserAssetsInNetworkAssets(
+      networks: allNetworks,
+      includingUserDeleted: false
+    )
+    typealias TokenNetworkAccounts = (token: BraveWallet.BlockchainToken, network: BraveWallet.NetworkInfo, accounts: [BraveWallet.AccountInfo])
+    let allTokenNetworkAccounts = allUserAssets.flatMap { networkAssets in
+      networkAssets.tokens.map { token in
+        TokenNetworkAccounts(
+          token: token,
+          network: networkAssets.network,
+          accounts: accounts.filter {
+            if token.coin == .fil {
+              return $0.keyringId == BraveWallet.KeyringId.keyringId(for: token.coin, on: token.chainId)
+            } else {
+              return $0.coin == token.coin
+            }
+          }
+        )
+      }
+    }
+    /// Fetch balance for each token, for all accounts applicable to that token
+    await withTaskGroup(
+      of: Void.self,
+      body: { @MainActor [rpcService] group in
+        for tokenNetworkAccounts in allTokenNetworkAccounts { // for each token
+          group.addTask { @MainActor in
+            let token = tokenNetworkAccounts.token
+            for account in tokenNetworkAccounts.accounts { // fetch balance for this token for each account
+              let balanceForToken = await rpcService.balance(
+                for: token,
+                in: account,
+                network: tokenNetworkAccounts.network
+              )
+              WalletUserAssetBalance.updateBalance(for: token, balance: String(balanceForToken ?? 0), account: account.address)
+            }
+          }
+        }
+      }
+    )
```

Comment:
I believe this can be simplified using extension on `BraveWalletJsonRpcService` (used in `SelectAccountTokenStore` & `AccountsStore`):
https://github.com/brave/brave-ios/blob/development/Sources/BraveWallet/Extensions/RpcServiceExtensions.swift#L298-L301

Slowly getting rid of these `TokenNetworkAccounts` typealiases and using the `NetworkAssets` directly 🙂.

Reply:
thanks. Have updated it in [56b6ac2](https://github.com/brave/brave-ios/pull/8726/commits/56b6ac2cb74d3d0950862cb078964c37179eac7b)
Please give a retest since balance refreshing is using another method now.